---
layout: single
published: true
title:  "[JAVA] 프로그램 오류, 예외 클래스의 계층구조 "
categories: Java
date: 2024-02-02 20:12:55
toc: true
toc_sticky: true
tag:   
  - Java
  - 자바
  - 프로그램 오류, 예외 클래스의 계층구조 
---

컴퓨터 하드웨어의 오작동 또는 고장으로 인해 응용프로그램 실행 오류가 발생하는 것을 에러(error)라고 합니다. 프로그램을 만들다보면 이런 에러 상황을 처리해야할 때가 있는데, 이것을 예외처리라고 합니다. 

에러 상황에서 프로그램이 강제 종료가 되는게 아니라, 에러 상황을 처리해줌으로써 프로그램이 정상 실행 상태가 유지되도록 할 수 있습니다. 프로그램이 오작동을 하지 않게 하기위한 자바의 배려라고 생각하시면 됩니다. 

## 예외는 언제 발생하는가

예외에 대하여 구체적으로 알아보기 전에, 어떤 상황에서 예외가 발생하는지 살펴봅시다. 

다음과 같이 존재하지 않는 파일을 열기 위한 자바 프로그램을 작성해봅시다.

```java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;

public class ExceptionSample {
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new FileReader("존재하지_않는_파일"));
		br.readLine();
		br.close();
	}
}
```

이 코드를 실행하면, 존재하지 않는 파일을 읽으려 하기 때문에, 원하는 파일을 찾을 수 없다는 오류가 발생합니다. 

`java.io.FileNotFoundException`

![image](https://github.com/BaxDailyGit/BaxDailyGit/assets/99312529/57a9ef72-3090-4973-b642-1e388f438e32)


또 다른 코드를 보겠습니다. 이번에는 0으로 다른 숫자를 나누는 프로그램을 만들어봅니다.

```java
public class DivideExceptionSample {
	public static void main(String[] args) {
		int result = 5 / 0;
	}
}
```

자바에서는 숫자를 0으로 나눌 수 없기 때문에 산술에 문제가 생겼다는 `ArithmeticException`이 발생합니다.

![image](https://github.com/BaxDailyGit/BaxDailyGit/assets/99312529/9fc3566b-a056-4a74-80f7-8d65024343e9)


마지막으로 한가지 예시를 더 보겠습니다. 

```java
public class ArrayExceptionSample {
	public static void main(String[] args) {
		int[] array = {3, 4, 5, 6};
		System.out.println(array[5]);
	}
}
```

코드를 보면 array 배열 크기가 4개인 숫자를 선언했는데, 인덱스 범위를 초과하여 사용할 경우 다음과 같은 오류가 발생합니다. 출력문에서 5번째 인덱스를 호출하고 있으니 당연히 오류가 발생하겠죠. `ArrayIndexOutOfBoundsException`이 발생합니다.

![image](https://github.com/BaxDailyGit/BaxDailyGit/assets/99312529/58a8e211-83fb-4967-9b03-00883c3d49d4)


자바는 이와 같이 예외가 발생하면 **프로그램을 중단하고 오류 메시지를 보여줍니다**. 만약 이 경우에 예외 처리를 해준다면, 프로그램 중단없이 오류 메시지를 출력하고 다음 프로세스로 넘어가도록 조치해줄 수 있겠죠. 이것이 바로 자바에서의 예외처리 입니다.

## 예외 클래스의 계층 구조

우선 예외 클래스에는 두가지 종류가 있습니다. 하나는 일반 예외(Exception), 다른 하나는 실행 예외(Runtime Exception)입니다. 

일반 예외는 컴파일러 체크 예외(Checked Exception) 라고도 하는데, 컴파일하는 과정에서 예외 처리 코드가 필요한지 검사하기 때문입니다. 만약 예외 처리 코드가 없다면 컴파일 오류가 발생합니다. 

실행 예외는 컴파일하는 과정에서 예외 처리 코드를 검사하지 않는 예외(Unchecked Exception)를 말합니다. 컴파일시 예외 처리를 확인하는 차이일 뿐, 두 가지 예외는 모두 예외 처리가 필요합니다. 

![image](https://github.com/BaxDailyGit/BaxDailyGit/assets/99312529/b47d46ec-79c9-4494-aac9-fcfb991c90e1)


- 모든 객체의 최상위 부모가 Object이듯, 예외도 최상위 부모는 Object
- Throwable : 최상위 예외.
- Error : 메모리 부족이나 심각한 시스템 오류같은 복구 불가능한 예외. 개발자가 잡을 수 없는 예외입니다.
- Exception : 컴파일러가 체크하는 checked exception. 단 RuntimeException은 예외.
- RuntimeException : 컴파일러가 체크하지 않는 unchecked exception, 런타임 예외

![image](https://github.com/BaxDailyGit/BaxDailyGit/assets/99312529/dd6e7a7c-9d78-4dec-88dc-e88a1187b27c)


![image](https://github.com/BaxDailyGit/BaxDailyGit/assets/99312529/33332f19-29d2-4c2e-83a3-3b814a79f7e2)
<span style="color:gray; display: block; text-align: right;">출처 : 이스트소프트 조성연</span>
